pkg/epp/flowcontrol/registry/registry_test.go

```diff
@@ -367,37 +367,41 @@ func TestFlowRegistry_GarbageCollection(t *testing.T) {
 		key := types.FlowKey{ID: "re-activated-flow", Priority: highPriority}
 		h.openConnectionOnFlow(key)
 
-		// Make the flow a candidate for GC.
-		h.fakeClock.Step(h.config.FlowGCTimeout + time.Second)
-		candidates := []types.FlowKey{key}
-
 		// Get the flow's state so we can manipulate its lock.
 		val, ok := h.fr.flowStates.Load(key)
 		require.True(t, ok, "Test setup: flow state must exist")
 		state := val.(*flowState)
 
-		// Lock the flow's `gcLock` to pause the GC.
+		// Acquire the lock before stepping the clock.
+		// This ensures the Background GC (which wakes up on Step) is blocked and cannot touch our flow until we release
+		// this lock.
 		state.gcLock.Lock()
 		defer state.gcLock.Unlock()
```

Comment:
Here you use `defer` to unlock the gcLock, yet in line 399 below, you explicitly unlock the gcLock.

That seems to be a bug here.

Reply:
It works, but it is a code smell.

This isn't included in the diff view, but look at the last line of this test case:

```
...
  h.assertFlowExists(key, "Flow should not be collected because it became active before the sweep")
  state.gcLock.Lock() // Re-lock for the deferred unlock.
})
```

It's actually a bit brittle. If the test were to fail (e.g., t.Fatal) after the explicit unlock but before the final re-lock, the defer would trigger a double-unlock panic, masking the actual test failure.

E.g., in

```
// Wait for the manual sweep to complete.
select {
case <-sweepDone:
  // Success
case <-time.After(time.Second):
  t.Fatal("verifyAndSweepFlows deadlocked or timed out")
}
```

Reply:
This test is necessarily a bit brittle and white-box as it deterministically triggers a race condition. I can clean this test up a bit by removing the `defer` entirely. However, the consequence of this is that if the test panics or fails between the lock and unlock, the lock will be leaked. This is probably fine for a unit test. Especially because the code in the critical section is unlikely to fail.

Revised, it looks like this:
```
...
state.gcLock.Lock()

h.fakeClock.Step(h.config.FlowGCTimeout + time.Second)
candidates := []types.FlowKey{key}

sweepDone := make(chan struct{})
go func() {
  defer close(sweepDone)
  h.fr.verifyAndSweepFlows(candidates)
}()

state.leaseCount.Add(1)
state.gcLock.Unlock()
...
```

Reply:
Let me know your thoughts - I think the revised code above is slightly preferable and less confusing for future maintainers.